Add Order model

ipantazi opened the pull request:
Added entity Order and OrderItem.

Added new classes: OrderController, OrderItemController, OrderDto, OrderItemServiceImpl, OrderServiceImpl.

Added new records: UpdateOrderDto, OrderItemDto, OrderRequestDto.

Added new interfaces: OrderItemMapper, OrderItemRepository, OrderItemService, OrderMapper, OrderRepository, OrderService.

Added changes to Liquibase.

Added method handleIllegalArgumentException() to CustomGlobalExceptionHandler, findShoppingCartByUserId() to ShoppingCartService.

Changed annotation size in field shippingAddress in record UserRegistrationRequestDto, class User.

ipantazi committed:
Added entity Order and OrderItem.

Added new classes: OrderController, OrderItemController, OrderDto, OrderItemServiceImpl, OrderServiceImpl.

Added new records: UpdateOrderDto, OrderItemDto, OrderRequestDto.

Added new interfaces: OrderItemMapper, OrderItemRepository, OrderItemService, OrderMapper, OrderRepository, OrderService.

Added changes to Liquibase.

Added method handleIllegalArgumentException() to CustomGlobalExceptionHandler, findShoppingCartByUserId() to ShoppingCartService.

Changed annotation size in field shippingAddress in record UserRegistrationRequestDto, class User.

Elena-Bruyako commented on src/main/java/onlinebookstore/controller/OrderItemController.java:
```diff
@@ -0,0 +1,50 @@
+@RestController
+@RequestMapping("/orders/{orderId}/items")
+@RequiredArgsConstructor
+public class OrderItemController {
```
all logic should be in OrderController

Elena-Bruyako commented on src/main/java/onlinebookstore/service/order/OrderServiceImpl.java:
```diff
@@ -0,0 +1,64 @@
+    private final ShoppingCartRepository shoppingCartRepository;
+
+    @Override
+    @Transactional
```
replace under the class

Elena-Bruyako commented on src/main/java/onlinebookstore/service/order/OrderServiceImpl.java:
```diff
@@ -0,0 +1,64 @@
+        ShoppingCart shoppingCart = shoppingCartService.findShoppingCartByUserId(userId);
+
+        if (shoppingCart.getCartItems().isEmpty()) {
+            throw new EntityNotFoundException("Shopping cart is empty");
```
```suggestion
            throw new OrderProcessingException("Shopping cart is empty");
```

Elena-Bruyako commented on src/main/java/onlinebookstore/service/order/OrderServiceImpl.java:
```diff
@@ -0,0 +1,64 @@
+        order.setShippingAddress(shippingAddress);
+        orderRepository.save(order);
+
+        if (order.getId() == null) {
```
remove

Elena-Bruyako reviewed (changes requested):
see comments

ipantazi committed:
All logic united in OrderController class and deleted OrderItemController class.

Added OrderProcessingException class and method handleOrderProcessingException() to CustomGlobalExceptionHandler.

Removed the check and replaced annotation @Transaction under the OrderServiceImpl class.

JJJazl commented on src/main/java/onlinebookstore/dto/user/UserRegistrationRequestDto.java:
```diff
@@ -39,7 +39,7 @@ public record UserRegistrationRequestDto(
         String lastName,
 
-        @Size(max = 255, message = "Shipping address must not exceed 255 characters.")
+        @Size(max = 100, message = "Shipping address must not exceed 255 characters.")
```
```suggestion
        @Size(max = 100, message = "Shipping address must not exceed 100 characters.")
```

JJJazl commented on src/main/java/onlinebookstore/model/OrderItem.java:
```diff
@@ -0,0 +1,45 @@
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
+    private Long id;
+
+    @ManyToOne(fetch = FetchType.LAZY, cascade = CascadeType.PERSIST)
```
I guess we don't need to use cascades on `OrderItem` side
```suggestion
    @ManyToOne(fetch = FetchType.LAZY, optional = false)
```

JJJazl commented on src/main/java/onlinebookstore/model/OrderItem.java:
```diff
@@ -0,0 +1,45 @@
+    @ToString.Exclude
+    private Order order;
+
+    @ManyToOne(fetch = FetchType.LAZY,optional = false)
```
```suggestion
    @ManyToOne(fetch = FetchType.LAZY, optional = false)
```

JJJazl commented on src/main/java/onlinebookstore/service/order/OrderServiceImpl.java:
```diff
@@ -0,0 +1,61 @@
+        ShoppingCart shoppingCart = shoppingCartService.findShoppingCartByUserId(userId);
+
+        if (shoppingCart.getCartItems().isEmpty()) {
+            throw new OrderProcessingException("Shopping cart is empty");
```
```suggestion
            throw new OrderProcessingException("Shopping cart is empty for user: " + userId);
```

JJJazl reviewed (approved):
Great! Left a few minor comments. Please, fix them before moving on to the next task

ipantazi committed:
Deleted cascade in class OrderItem.
